File: services/monitoring/sync_monitor.js
/**
 * Synchronization Monitoring Service
 * 
 * This module provides monitoring capabilities for the Canvas-Discourse synchronization,
 * allowing administrators to view sync status, error rates, and sync history.
 */

class SyncMonitor {
  /**
   * Create a new synchronization monitoring service
   * 
   * @param {Object} options - Configuration options
   * @param {SyncState} options.syncState - Reference to the sync state manager
   * @param {SyncTransaction} options.syncTransaction - Reference to transaction manager
   */
  constructor({ syncState, syncTransaction }) {
    this.syncState = syncState;
    this.syncTransaction = syncTransaction;
    this.cachedStats = null;
    this.cacheTime = null;
    this.CACHE_TTL = 60000; // 1 minute cache
  }

  /**
   * Get overall synchronization statistics
   * 
   * @returns {Promise<Object>} Statistics about synchronization
   */
  async getStatistics() {
    // Use cached stats if available and recent
    if (this.cachedStats && (Date.now() - this.cacheTime < this.CACHE_TTL)) {
      return this.cachedStats;
    }

    const stats = {
      entities: {
        users: await this._getEntityStats('user'),
        courses: await this._getEntityStats('course'),
        assignments: await this._getEntityStats('assignment'),
        discussions: await this._getEntityStats('discussion'),
        posts: await this._getEntityStats('post')
      },
      transactions: await this._getTransactionStats(),
      overall: {
        lastSync: await this._getLastSyncTime(),
        errorRate: await this._calculateErrorRate()
      }
    };

    // Cache the results
    this.cachedStats = stats;
    this.cacheTime = Date.now();

    return stats;
  }

  /**
   * Get pending synchronization items that need attention
   * 
   * @param {number} limit - Maximum number of items to return
   * @returns {Promise<Array>} List of sync items needing attention
   */
  async getPendingItems(limit = 100) {
    return await this.syncState.getFailedSyncs(limit);
  }

  /**
   * Get synchronization history for a specific entity
   * 
   * @param {string} entityType - Type of entity (user, course, etc.)
   * @param {string} entityId - ID of the entity
   * @returns {Promise<Array>} Synchronization history for the entity
   */
  async getEntitySyncHistory(entityType, entityId) {
    const transactions = await this.syncTransaction.getTransactionsForEntity(entityType, entityId);
    const syncState = await this.syncState.getSyncStatus(entityType, entityId);
    
    return {
      currentStatus: syncState,
      syncHistory: transactions.map(t => ({
        id: t.id,
        timestamp: t.timestamp,
        operation: t.operation,
        status: t.status,
        duration: t.endTime ? (t.endTime - t.startTime) : null,
        error: t.error
      }))
    };
  }

  /**
   * Trigger a manual resync for a failing entity
   * 
   * @param {string} entityType - Type of entity to resync
   * @param {string} entityId - ID of the entity to resync
   * @param {string} priority - Priority for the resync operation
   * @returns {Promise<boolean>} True if resync was successfully triggered
   */
  async triggerResync(entityType, entityId, priority = 'high') {
    // Reset the sync state
    await this.syncState.resetSyncState(entityType, entityId);
    
    // Create a new sync transaction
    const transaction = await this.syncTransaction.beginTransaction({
      entityType,
      entityId,
      operation: 'MANUAL_RESYNC',
      initiatedBy: 'admin-dashboard'
    });
    
    // Publish the event to the sync service
    // Note: In a real implementation, this would call the actual sync service
    console.log(`Publishing manual resync event for ${entityType}:${entityId}`);
    
    await this.syncTransaction.recordStep(transaction.id, 'QUEUED', 'Manual resync requested');
    
    return true;
  }

  // Private helper methods
  
  /**
   * Get statistics for a specific entity type
   * 
   * @private
   * @param {string} entityType - Type of entity
   * @returns {Promise<Object>} Statistics for the entity type
   */
  async _getEntityStats(entityType) {
    const counts = await this.syncState.getSyncCountsByStatus(entityType);
    const totalCount = Object.values(counts).reduce((sum, count) => sum + count, 0);
    
    return {
      total: totalCount,
      synced: counts.SYNCED || 0,
      pending: counts.PENDING || 0,
      failed: counts.FAILED || 0,
      inProgress: counts.IN_PROGRESS || 0,
      syncPercentage: totalCount > 0 ? Math.round((counts.SYNCED || 0) / totalCount * 100) : 0
    };
  }

  /**
   * Get statistics about transactions
   * 
   * @private
   * @returns {Promise<Object>} Transaction statistics
   */
  async _getTransactionStats() {
    const last24Hours = new Date(Date.now() - 24 * 60 * 60 * 1000);
    const transactions = await this.syncTransaction.getTransactions({ after: last24Hours });
    
    const stats = {
      last24Hours: transactions.length,
      successful: transactions.filter(t => t.status === 'COMPLETED').length,
      failed: transactions.filter(t => t.status === 'FAILED').length,
      inProgress: transactions.filter(t => t.status === 'IN_PROGRESS').length
    };
    
    stats.successRate = stats.last24Hours > 0 ? 
      Math.round((stats.successful / stats.last24Hours) * 100) : 0;
    
    return stats;
  }

  /**
   * Get the timestamp of the last successful sync
   * 
   * @private
   * @returns {Promise<string>} ISO timestamp of last sync
   */
  async _getLastSyncTime() {
    const lastTransaction = await this.syncTransaction.getLatestSuccessfulTransaction();
    return lastTransaction ? lastTransaction.endTime : null;
  }

  /**
   * Calculate the current error rate
   * 
   * @private
   * @returns {Promise<number>} Error rate as a percentage
   */
  async _calculateErrorRate() {
    const last100 = await this.syncTransaction.getTransactions({ limit: 100 });
    if (last100.length === 0) return 0;
    
    const failedCount = last100.filter(t => t.status === 'FAILED').length;
    return Math.round((failedCount / last100.length) * 100);
  }
}

module.exports = { SyncMonitor };

File: services/monitoring/sync_monitor.rs
use crate::services::integration::sync_state::SyncState;
use crate::services::integration::sync_transaction::SyncTransaction;
use serde::{Serialize, Deserialize};
use std::sync::Arc;
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
use tokio::sync::RwLock;
use thiserror::Error;

/// Error types for the sync monitor
#[derive(Error, Debug)]
pub enum SyncMonitorError {
    #[error("Failed to get entity statistics: {0}")]
    EntityStatsError(String),
    
    #[error("Failed to get transaction statistics: {0}")]
    TransactionStatsError(String),
    
    #[error("Failed to get sync status: {0}")]
    SyncStatusError(String),
    
    #[error("Invalid entity type: {0}")]
    InvalidEntityType(String),
    
    #[error("Entity not found: {0} {1}")]
    EntityNotFound(String, String),
}

/// Entity synchronization statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityStats {
    pub total: usize,
    pub synced: usize,
    pub failed: usize,
    pub pending: usize,
    pub success_rate: f64,
}

/// Transaction statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionStats {
    pub total: usize,
    pub success: usize,
    pub failed: usize,
    pub in_progress: usize,
    pub average_duration_ms: f64,
}

/// Overall synchronization statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncStats {
    pub entities: EntityTypeStats,
    pub transactions: TransactionStats,
    pub overall: OverallStats,
}

/// Entity type statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityTypeStats {
    pub users: EntityStats,
    pub courses: EntityStats,
    pub assignments: EntityStats,
    pub discussions: EntityStats,
    pub posts: EntityStats,
}

/// Overall synchronization metrics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OverallStats {
    pub last_sync: Option<u64>,
    pub error_rate: f64,
}

/// Entity synchronization history
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntitySyncHistory {
    pub current_status: serde_json::Value,
    pub sync_history: Vec<SyncHistoryItem>,
}

/// Individual sync history item
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncHistoryItem {
    pub id: String,
    pub timestamp: u64,
    pub operation: String,
    pub status: String,
    pub duration: Option<u64>,
    pub error: Option<String>,
}

/// Synchronization Monitor Service
pub struct SyncMonitor {
    sync_state: Arc<SyncState>,
    sync_transaction: Arc<SyncTransaction>,
    cached_stats: RwLock<Option<(SyncStats, Instant)>>,
    cache_ttl: Duration,
}

impl SyncMonitor {
    /// Create a new synchronization monitoring service
    pub fn new(sync_state: Arc<SyncState>, sync_transaction: Arc<SyncTransaction>) -> Self {
        SyncMonitor {
            sync_state,
            sync_transaction,
            cached_stats: RwLock::new(None),
            cache_ttl: Duration::from_secs(60), // 1 minute cache
        }
    }

    /// Get overall synchronization statistics
    pub async fn get_statistics(&self) -> Result<SyncStats, SyncMonitorError> {
        // Check cache first
        {
            let cached = self.cached_stats.read().await;
            if let Some((stats, cache_time)) = &*cached {
                if cache_time.elapsed() < self.cache_ttl {
                    return Ok(stats.clone());
                }
            }
        }

        // Cache miss or expired, compute fresh stats
        let users = self.get_entity_stats("user").await?;
        let courses = self.get_entity_stats("course").await?;
        let assignments = self.get_entity_stats("assignment").await?;
        let discussions = self.get_entity_stats("discussion").await?;
        let posts = self.get_entity_stats("post").await?;
        
        let entity_stats = EntityTypeStats {
            users,
            courses,
            assignments,
            discussions,
            posts,
        };
        
        let transaction_stats = self.get_transaction_stats().await?;
        
        let last_sync = self.get_last_sync_time().await?;
        let error_rate = self.calculate_error_rate().await?;
        
        let overall = OverallStats {
            last_sync,
            error_rate,
        };
        
        let stats = SyncStats {
            entities: entity_stats,
            transactions: transaction_stats,
            overall,
        };
        
        // Update cache
        let mut cache = self.cached_stats.write().await;
        *cache = Some((stats.clone(), Instant::now()));
        
        Ok(stats)
    }

    /// Get pending synchronization items that need attention
    pub async fn get_pending_items(&self, limit: usize) -> Result<Vec<serde_json::Value>, SyncMonitorError> {
        self.sync_state.get_failed_syncs(limit)
            .await
            .map_err(|e| SyncMonitorError::SyncStatusError(e.to_string()))
    }

    /// Get synchronization history for a specific entity
    pub async fn get_entity_sync_history(
        &self,
        entity_type: &str,
        entity_id: &str,
    ) -> Result<EntitySyncHistory, SyncMonitorError> {
        // Get transaction history
        let transactions = self.sync_transaction.get_transactions_for_entity(entity_type, entity_id)
            .await
            .map_err(|e| SyncMonitorError::TransactionStatsError(e.to_string()))?;
        
        // Get current sync state
        let sync_state = self.sync_state.get_sync_status(entity_type, entity_id)
            .await
            .map_err(|e| SyncMonitorError::SyncStatusError(e.to_string()))?;
        
        // Convert transactions to history items
        let sync_history = transactions.into_iter()
            .map(|t| {
                let duration = if let (Some(start), Some(end)) = (t.start_time, t.end_time) {
                    Some(end - start)
                } else {
                    None
                };
                
                SyncHistoryItem {
                    id: t.id,
                    timestamp: t.timestamp,
                    operation: t.operation,
                    status: t.status,
                    duration,
                    error: t.error,
                }
            })
            .collect();
        
        Ok(EntitySyncHistory {
            current_status: sync_state,
            sync_history,
        })
    }

    /// Trigger a manual resync for a failing entity
    pub async fn trigger_resync(
        &self,
        entity_type: &str,
        entity_id: &str,
        priority: Option<&str>,
    ) -> Result<bool, SyncMonitorError> {
        // Validate entity type
        match entity_type {
            "user" | "course" | "assignment" | "discussion" | "post" => {},
            _ => return Err(SyncMonitorError::InvalidEntityType(entity_type.to_string())),
        }
        
        // Reset the sync state
        self.sync_state.reset_sync_state(entity_type, entity_id)
            .await
            .map_err(|e| SyncMonitorError::SyncStatusError(e.to_string()))?;
        
        // Create a new sync transaction
        let priority = priority.unwrap_or("high").to_string();
        
        self.sync_transaction.create_transaction(entity_type, entity_id, "resync", Some(priority))
            .await
            .map_err(|e| SyncMonitorError::TransactionStatsError(e.to_string()))?;
        
        Ok(true)
    }

    // Private helper methods
    
    /// Get statistics for a specific entity type
    async fn get_entity_stats(&self, entity_type: &str) -> Result<EntityStats, SyncMonitorError> {
        let stats = self.sync_state.get_entity_stats(entity_type)
            .await
            .map_err(|e| SyncMonitorError::EntityStatsError(e.to_string()))?;
        
        let total = stats["total"].as_u64().unwrap_or(0) as usize;
        let synced = stats["synced"].as_u64().unwrap_or(0) as usize;
        let failed = stats["failed"].as_u64().unwrap_or(0) as usize;
        let pending = stats["pending"].as_u64().unwrap_or(0) as usize;
        
        let success_rate = if total > 0 {
            synced as f64 / total as f64
        } else {
            0.0
        };
        
        Ok(EntityStats {
            total,
            synced,
            failed,
            pending,
            success_rate,
        })
    }

    /// Get statistics about transactions
    async fn get_transaction_stats(&self) -> Result<TransactionStats, SyncMonitorError> {
        let stats = self.sync_transaction.get_transaction_stats()
            .await
            .map_err(|e| SyncMonitorError::TransactionStatsError(e.to_string()))?;
        
        let total = stats["total"].as_u64().unwrap_or(0) as usize;
        let success = stats["success"].as_u64().unwrap_or(0) as usize;
        let failed = stats["failed"].as_u64().unwrap_or(0) as usize;
        let in_progress = stats["in_progress"].as_u64().unwrap_or(0) as usize;
        let average_duration_ms = stats["average_duration_ms"].as_f64().unwrap_or(0.0);
        
        Ok(TransactionStats {
            total,
            success,
            failed,
            in_progress,
            average_duration_ms,
        })
    }

    /// Get the timestamp of the last successful sync
    async fn get_last_sync_time(&self) -> Result<Option<u64>, SyncMonitorError> {
        let last_sync = self.sync_transaction.get_last_successful_sync()
            .await
            .map_err(|e| SyncMonitorError::TransactionStatsError(e.to_string()))?;
        
        Ok(last_sync)
    }

    /// Calculate the overall error rate in synchronization
    async fn calculate_error_rate(&self) -> Result<f64, SyncMonitorError> {
        let stats = self.sync_transaction.get_transaction_stats()
            .await
            .map_err(|e| SyncMonitorError::TransactionStatsError(e.to_string()))?;
        
        let total = stats["total"].as_u64().unwrap_or(0) as f64;
        let failed = stats["failed"].as_u64().unwrap_or(0) as f64;
        
        if total > 0.0 {
            Ok(failed / total)
        } else {
            Ok(0.0)
        }
    }
}
